# Product Recommender for Online Stores

How online stores can help customers to buy the right product from their millions of products,so that online retailers can retain the customer from going elsewhere and see yearly increase in their average order value. Build a recommendation system so that customers can get the right product with personalized info.

Online stores show millions of products to the customer from their catalog. Choosing the correct product for their needs is becoming difficult because of so much information.Since customers are more likely to buy based on personalized recommendations, management has decided to go for a recommendation system so that they can retain the customer and hence increase yearly product sales

## Refernces

The dataset here is taken from the below website.

Source - Amazon Reviews data (http://jmcauley.ucsd.edu/data/amazon/links.html) The repository has several datasets. For this case study, I am using the Electronics dataset.

This dataset contains product reviews and metadata from Amazon,reviews spanning May 1996 - July 2014.

Files are meta_Books.json.gz and reviews_Electronics.json.gz. DAta from these fiels are parsed to dataframe and saved as csv to local drive.Here is the code to panada data frame - 

import pandas as pd
import gzip

def parse(path):
  g = gzip.open(path, 'rb')
  for l in g:
    yield eval(l)

def getDF(path):
  i = 0
  df = {}
  for d in parse(path):
    df[i] = d
    i += 1
  return pd.DataFrame.from_dict(df, orient='index')

df = getDF('reviews_Video_Games.json.gz')


This dataset includes reviews (ratings, text, helpfulness votes), product metadata (descriptions, category information, price, brand, and image features), and links (also viewed/also bought graphs).

## Data Dictionary

### Sample review dataset:

{
  "reviewerID": "A2SUAM1J3GNN3B",
  "asin": "0000013714",
  "reviewerName": "J. McDonald",
  "helpful": [2, 3],
  "reviewText": "I bought this for my husband who plays the piano.  He is having a wonderful time playing these old hymns.  The music  is at times hard to read because we think the book was published for singing from more than playing from.  Great purchase though!",
  "overall": 5.0,
  "summary": "Heavenly Highway Hymns",
  "unixReviewTime": 1252800000,
  "reviewTime": "09 13, 2009"
}

where

* reviewerID - ID of the reviewer, e.g. A2SUAM1J3GNN3B
* asin - ID of the product, e.g. 0000013714
* reviewerName - name of the reviewer
* helpful - helpfulness rating of the review, e.g. 2/3
* reviewText - text of the review
* overall - rating of the product
* summary - summary of the review
* unixReviewTime - time of the review (unix time)
* reviewTime - time of the review (raw)

## Sample Metadata

{
  "asin": "0000031852",
  "title": "Girls Ballet Tutu Zebra Hot Pink",
  "price": 3.17,
  "imUrl": "http://ecx.images-amazon.com/images/I/51fAmVkTbyL._SY300_.jpg",
  "related":
  {
    "also_bought": ["B00JHONN1S", "B002BZX8Z6", "B00D2K1M3O", "0000031909", "B00613WDTQ", "B00D0WDS9A", "B00D0GCI8S", "0000031895", "B003AVKOP2", "B003AVEU6G", "B003IEDM9Q", "B002R0FA24", "B00D23MC6W", "B00D2K0PA0", "B00538F5OK", "B00CEV86I6", "B002R0FABA", "B00D10CLVW", "B003AVNY6I", "B002GZGI4E", "B001T9NUFS", "B002R0F7FE", "B00E1YRI4C", "B008UBQZKU", "B00D103F8U", "B007R2RM8W"],
    "also_viewed": ["B002BZX8Z6", "B00JHONN1S", "B008F0SU0Y", "B00D23MC6W", "B00AFDOPDA", "B00E1YRI4C", "B002GZGI4E", "B003AVKOP2", "B00D9C1WBM", "B00CEV8366", "B00CEUX0D8", "B0079ME3KU", "B00CEUWY8K", "B004FOEEHC", "0000031895", "B00BC4GY9Y", "B003XRKA7A", "B00K18LKX2", "B00EM7KAG6", "B00AMQ17JA", "B00D9C32NI", "B002C3Y6WG", "B00JLL4L5Y", "B003AVNY6I", "B008UBQZKU", "B00D0WDS9A", "B00613WDTQ", "B00538F5OK", "B005C4Y4F6", "B004LHZ1NY", "B00CPHX76U", "B00CEUWUZC", "B00IJVASUE", "B00GOR07RE", "B00J2GTM0W", "B00JHNSNSM", "B003IEDM9Q", "B00CYBU84G", "B008VV8NSQ", "B00CYBULSO", "B00I2UHSZA", "B005F50FXC", "B007LCQI3S", "B00DP68AVW", "B009RXWNSI", "B003AVEU6G", "B00HSOJB9M", "B00EHAGZNA", "B0046W9T8C", "B00E79VW6Q", "B00D10CLVW", "B00B0AVO54", "B00E95LC8Q", "B00GOR92SO", "B007ZN5Y56", "B00AL2569W", "B00B608000", "B008F0SMUC", "B00BFXLZ8M"],
    "bought_together": ["B002BZX8Z6"]
  },
  "salesRank": {"Toys & Games": 211836},
  "brand": "Coxlures",
  "categories": [["Sports & Outdoors", "Other Sports", "Dance"]]
}

where

* asin - ID of the product, e.g. 0000031852
* title - name of the product
* price - price in US dollars (at time of crawl)
* imUrl - url of the product image
* related - related products (also bought, also viewed, bought together, buy after viewing)
* salesRank - sales rank information
* brand - brand name
* categories - list of categories the product belongs to

The data is being too huge and is time consuming to load and further process, few rows are droppped based on the null creteria and saved to new CSV named "2.reviews_cleaned.csv" to local drive.
And this new file is used for all future DSM steps. 

There is new notebook is used for further data wrangling and EDA and is named "Data Wrangling and EDA".

## Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Loading reviews data from CSV
dfr = pd.read_csv('../Data/reviews_Electronics.csv',header=0)

In [3]:
dfr.head()

reviewerID        asin          reviewerName helpful  \
0   AKM1MP6P0OYPR  0132793040  Vicki Gibson "momo4"  [1, 1]   
1  A2CX7LUOHB2NDG  0321732944                Bernie  [0, 0]   
2  A2NWSAGRHCP8N5  0439886341           bowmans2007  [1, 1]   
3  A2WNBOD3WNDNKT  0439886341                   JAL  [1, 1]   
4  A1GI0U4ZRJA8WN  0439886341             Truthfull  [4, 4]   

                                          reviewText  overall  \
0  Corey Barker does a great job of explaining Bl...      5.0   
1  While many beginner DVDs try to teach you ever...      5.0   
2  It never worked. My daughter worked to earn th...      1.0   
3  Some of the functions did not work properly.  ...      3.0   
4  Do not waste your money on this thing it is te...      1.0   

                                             summary  unixReviewTime  \
0                                      Very thorough      1365811200   
1  Adobe Photoshop CS5 Crash Course with master P...      1341100800   
2                                absolutely horrible      1367193600   
3                                      Disappointing      1374451200   
4                     TERRIBLE DONT WASTE YOUR MONEY      1334707200   

    reviewTime  
0  04 13, 2013  
1   07 1, 2012  
2  04 29, 2013  
3  07 22, 2013  
4  04 18, 2012

In [4]:
dfr.shape

(7824482, 9)

In [5]:
dfr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7824482 entries, 0 to 7824481
Data columns (total 9 columns):
 #   Column          Dtype  
---  ------          -----  
 0   reviewerID      object 
 1   asin            object 
 2   reviewerName    object 
 3   helpful         object 
 4   reviewText      object 
 5   overall         float64
 6   summary         object 
 7   unixReviewTime  int64  
 8   reviewTime      object 
dtypes: float64(1), int64(1), object(7)
memory usage: 537.3+ MB


There are 7824482 records in the reviews dataset. 

In [6]:
#Loading Meta data from CSV
dfm = pd.read_csv('../Data/meta_Electronics.csv',header=0)

In [7]:
#merge reviewes and meta
dfrm0 = pd.merge(dfr,dfm,on='asin',how='left')

In [8]:
dfrm0.head()

reviewerID        asin          reviewerName helpful  \
0   AKM1MP6P0OYPR  0132793040  Vicki Gibson "momo4"  [1, 1]   
1  A2CX7LUOHB2NDG  0321732944                Bernie  [0, 0]   
2  A2NWSAGRHCP8N5  0439886341           bowmans2007  [1, 1]   
3  A2WNBOD3WNDNKT  0439886341                   JAL  [1, 1]   
4  A1GI0U4ZRJA8WN  0439886341             Truthfull  [4, 4]   

                                          reviewText  overall  \
0  Corey Barker does a great job of explaining Bl...      5.0   
1  While many beginner DVDs try to teach you ever...      5.0   
2  It never worked. My daughter worked to earn th...      1.0   
3  Some of the functions did not work properly.  ...      3.0   
4  Do not waste your money on this thing it is te...      1.0   

                                             summary  unixReviewTime  \
0                                      Very thorough      1365811200   
1  Adobe Photoshop CS5 Crash Course with master P...      1341100800   
2                                absolutely horrible      1367193600   
3                                      Disappointing      1374451200   
4                     TERRIBLE DONT WASTE YOUR MONEY      1334707200   

    reviewTime                                              imUrl  \
0  04 13, 2013  http://ecx.images-amazon.com/images/I/31JIPhp%...   
1   07 1, 2012  http://ecx.images-amazon.com/images/I/31uogm6Y...   
2  04 29, 2013  http://ecx.images-amazon.com/images/I/51k0qa8f...   
3  07 22, 2013  http://ecx.images-amazon.com/images/I/51k0qa8f...   
4  04 18, 2012  http://ecx.images-amazon.com/images/I/51k0qa8f...   

                                         description  \
0  The Kelby Training DVD Mastering Blend Modes i...   
1                                                NaN   
2                    Digital Organizer and Messenger   
3                    Digital Organizer and Messenger   
4                    Digital Organizer and Messenger   

                                          categories  \
0  [['Electronics', 'Computers & Accessories', 'C...   
1  [['Electronics', 'Computers & Accessories', 'C...   
2  [['Electronics', 'Computers & Accessories', 'P...   
3  [['Electronics', 'Computers & Accessories', 'P...   
4  [['Electronics', 'Computers & Accessories', 'P...   

                                               title  price  \
0  Kelby Training DVD: Mastering Blend Modes in A...    NaN   
1  Kelby Training DVD: Adobe Photoshop CS5 Crash ...    NaN   
2                    Digital Organizer and Messenger   8.15   
3                    Digital Organizer and Messenger   8.15   
4                    Digital Organizer and Messenger   8.15   

                 salesRank                                            related  \
0                      NaN                                                NaN   
1                      NaN                                                NaN   
2  {'Electronics': 144944}  {'also_viewed': ['0545016266', 'B009ECM8QY', '...   
3  {'Electronics': 144944}  {'also_viewed': ['0545016266', 'B009ECM8QY', '...   
4  {'Electronics': 144944}  {'also_viewed': ['0545016266', 'B009ECM8QY', '...   

  brand  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

In [9]:
dfrm0.shape

(7824482, 17)

In [10]:
dfrm0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7824482 entries, 0 to 7824481
Data columns (total 17 columns):
 #   Column          Dtype  
---  ------          -----  
 0   reviewerID      object 
 1   asin            object 
 2   reviewerName    object 
 3   helpful         object 
 4   reviewText      object 
 5   overall         float64
 6   summary         object 
 7   unixReviewTime  int64  
 8   reviewTime      object 
 9   imUrl           object 
 10  description     object 
 11  categories      object 
 12  title           object 
 13  price           float64
 14  salesRank       object 
 15  related         object 
 16  brand           object 
dtypes: float64(2), int64(1), object(14)
memory usage: 1.0+ GB


In [11]:
dfrm0.isnull().sum()

reviewerID              0
asin                    0
reviewerName        99675
helpful                 0
reviewText           3719
overall                 0
summary               106
unixReviewTime          0
reviewTime              0
imUrl                5027
description       1685926
categories              0
title              293559
price              472727
salesRank         4541735
related            375261
brand             4071712
dtype: int64

In [12]:
dfrm=dfrm0.dropna(subset=['brand','title','salesRank','reviewerName','reviewText','description','price','related'])

In [13]:
dfrm.isnull().sum()

reviewerID         0
asin               0
reviewerName       0
helpful            0
reviewText         0
overall            0
summary           21
unixReviewTime     0
reviewTime         0
imUrl              0
description        0
categories         0
title              0
price              0
salesRank          0
related            0
brand              0
dtype: int64

In [14]:
# Checking for nunique value in the data attributes
dfrm.nunique().sort_values(ascending = False)

reviewText        1362754
reviewerID        1036910
summary            815559
reviewerName       787999
asin                35192
related             35102
title               34988
description         31251
imUrl               27561
salesRank           24073
price                8732
reviewTime           4886
unixReviewTime       4886
helpful              4884
brand                4733
categories           1001
overall                 5
dtype: int64

In [15]:
dfrm.head()

reviewerID        asin   reviewerName helpful  \
76  A2B4F6LP0ZLI8E  0594296420          Carol  [1, 1]   
77  A1RKH6803HVOQL  0594296420      Col "Col"  [1, 1]   
78   A3ZNVCQCG57AY  0594296420       Kelly K.  [0, 0]   
79  A1HM74IXWXQ5D8  0594296420  Matthew Frank  [1, 1]   
80  A2DUI0YVE1DZ48  0594296420       Pen Name  [0, 0]   

                                           reviewText  overall  \
76  I bought this item to replace the original cha...      3.0   
77  This is a NOOK product and works as advertised...      5.0   
78  I purchased this plug to see if it would work ...      5.0   
79  Got a nook for my wife and figured a wall adap...      5.0   
80  Perfect for me so that I don't have to take my...      5.0   

                                     summary  unixReviewTime   reviewTime  \
76                   Plug replacement needed      1398556800  04 27, 2014   
77             Works with NOOK as advertised      1365379200   04 8, 2013   
78  Works to Charge my Toshiba Excite Tablet      1359331200  01 28, 2013   
79                                Great Item      1392854400  02 20, 2014   
80                     Nook charging adapter      1388016000  12 26, 2013   

                                                imUrl  \
76  http://ecx.images-amazon.com/images/I/41Wwbxms...   
77  http://ecx.images-amazon.com/images/I/41Wwbxms...   
78  http://ecx.images-amazon.com/images/I/41Wwbxms...   
79  http://ecx.images-amazon.com/images/I/41Wwbxms...   
80  http://ecx.images-amazon.com/images/I/41Wwbxms...   

                                          description  \
76  If you prefer to charge your NOOK Simple Touch...   
77  If you prefer to charge your NOOK Simple Touch...   
78  If you prefer to charge your NOOK Simple Touch...   
79  If you prefer to charge your NOOK Simple Touch...   
80  If you prefer to charge your NOOK Simple Touch...   

                                           categories  \
76  [['Electronics', 'Computers & Accessories', 'L...   
77  [['Electronics', 'Computers & Accessories', 'L...   
78  [['Electronics', 'Computers & Accessories', 'L...   
79  [['Electronics', 'Computers & Accessories', 'L...   
80  [['Electronics', 'Computers & Accessories', 'L...   

                                                title  price salesRank  \
76  Barnes &amp; Noble NOOK Simple Touch Wall Adap...   6.99        {}   
77  Barnes &amp; Noble NOOK Simple Touch Wall Adap...   6.99        {}   
78  Barnes &amp; Noble NOOK Simple Touch Wall Adap...   6.99        {}   
79  Barnes &amp; Noble NOOK Simple Touch Wall Adap...   6.99        {}   
80  Barnes &amp; Noble NOOK Simple Touch Wall Adap...   6.99        {}   

                                              related               brand  
76  {'also_bought': ['B0087UBS4M', 'B0032UP32K', '...  Barnes &amp; Noble  
77  {'also_bought': ['B0087UBS4M', 'B0032UP32K', '...  Barnes &amp; Noble  
78  {'also_bought': ['B0087UBS4M', 'B0032UP32K', '...  Barnes &amp; Noble  
79  {'also_bought': ['B0087UBS4M', 'B0032UP32K', '...  Barnes &amp; Noble  
80  {'also_bought': ['B0087UBS4M', 'B0032UP32K', '...  Barnes &amp; Noble

In [16]:
dfrm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1365152 entries, 76 to 7824399
Data columns (total 17 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   reviewerID      1365152 non-null  object 
 1   asin            1365152 non-null  object 
 2   reviewerName    1365152 non-null  object 
 3   helpful         1365152 non-null  object 
 4   reviewText      1365152 non-null  object 
 5   overall         1365152 non-null  float64
 6   summary         1365131 non-null  object 
 7   unixReviewTime  1365152 non-null  int64  
 8   reviewTime      1365152 non-null  object 
 9   imUrl           1365152 non-null  object 
 10  description     1365152 non-null  object 
 11  categories      1365152 non-null  object 
 12  title           1365152 non-null  object 
 13  price           1365152 non-null  float64
 14  salesRank       1365152 non-null  object 
 15  related         1365152 non-null  object 
 16  brand           1365152 non-null  o

In [19]:
# save the data to a new csv file
import os
datapath = '../data'

datapath= os.path.join(datapath, '2.reviews_cleaned.csv')
#if not os.path.exists(datapath_ccattr_step2):
dfrm.to_csv(datapath, index=False)

As dataset is too big, it is now reduced to 1365152 records. Now on these data is used for further DSM steps and Wrangling and EDA continues in new notebook "Data Wrangling and EDA". 